presto-main/src/main/java/com/facebook/presto/sql/planner/LambdaCaptureDesugaringRewriter.java

```diff
@@ -0,0 +1,139 @@
+import com.facebook.presto.sql.tree.BindExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.LambdaArgumentDeclaration;
+import com.facebook.presto.sql.tree.LambdaExpression;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class LambdaCaptureDesugaringRewriter
+{
+    private final Map<Symbol, Type> symbolTypes;
+    private final SymbolAllocator symbolAllocator;
+
+    public LambdaCaptureDesugaringRewriter(Map<Symbol, Type> symbolTypes, SymbolAllocator symbolAllocator)
+    {
+        this.symbolTypes = requireNonNull(symbolTypes, "symbolTypes is null");
+        this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+    }
+
+    public Expression rewrite(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression, new Context());
+    }
+
+    private static Expression replaceSymbols(Expression expression, ImmutableMap<Symbol, Symbol> symbolMapping)
+    {
+        return ExpressionTreeRewriter.rewriteWith(
+                new ExpressionRewriter<Void>() {
+                    @Override
+                    public Expression rewriteSymbolReference(SymbolReference node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+                    {
+                        Symbol mapTo = symbolMapping.get(new Symbol(node.getName()));
+                        if (mapTo == null) {
+                            return node;
+                        }
+                        return mapTo.toSymbolReference();
+                    }
+                },
+                expression);
+    }
+
+    public class Visitor
+            extends ExpressionRewriter<Context>
+    {
+        @Override
+        public Expression rewriteLambdaExpression(LambdaExpression node, Context context, ExpressionTreeRewriter<Context> treeRewriter)
+        {
+            HashSet<Symbol> referencedSymbols = new HashSet<>();
+            Expression rewrittenBody = treeRewriter.rewrite(node.getBody(), context.withReferencedSymbols(referencedSymbols));
+
+            ImmutableList<Symbol> lambdaArguments = node.getArguments().stream()
+                    .map(LambdaArgumentDeclaration::getName)
+                    .map(Symbol::new)
+                    .collect(toImmutableList());
+            referencedSymbols.removeAll(lambdaArguments);
+
+            // x -> f(x, captureSymbol)    will be rewritten into
+            // "$internal$bind"(captureSymbol, (extraSymbol, x) -> f(x, extraSymbol))
```

Comment:
The type of `bind` is not `X, (X, Y) -> Z => (Y) ->Z`.

There is no way to write out the type of `bind`. It has a type that looks like:
* X, (X) -> Y => () -> Y
* X1, (X1, X2) -> Y => (X1, X2) -> Y
* X1, (X1, X2, X3) -> Y => (X1, X2, X3) -> Y
* X1, (X1, X2, X3, X4) -> Y => (X1, X2, X3, X4) -> Y
* ...

The above list is somewhat similar to the type of function application (i.e. the `()` operator as in `f(x)`): 
* X, (X) -> Y => Y
* X1, X2, (X1, X2) -> Y => Y
* X1, X2, X3, (X1, X2, X3) -> Y => Y
* X1, X2, X3, X4, (X1, X2, X3, X4) -> Y => Y
* ...

Now, in Presto, functions defined in Java is not treated the same way as lambda functions. But I believe you got the idea.

Bind is fundamentally first class. Just like you wouldn't want to give a type to function application, or treat function application as a normal function, the same cannot be done on bind.